quinn-proto/src/connection/streams/state.rs

```diff
@@ -1553,4 +1581,77 @@ mod tests {
             assert_eq!(client.max_remote[Dir::Bi as usize], 201);
         }
     }
+
+    #[test]
+    fn expand_receive_window() {
+        let mut server = make(Side::Server);
+        let new_receive_window = 2 * 1024 * 1024u32;
+        let explanded = server.set_receive_window(new_receive_window.into());
+        assert!(explanded);
+        assert_eq!(server.receive_window, new_receive_window as u64);
+        assert_eq!(server.local_max_data, new_receive_window as u64);
+        assert_eq!(server.receive_window_shrink_debt, 0);
+        let prev_local_max_data = server.local_max_data;
+
+        // credit, expecting all of them added to local_max_data
+        let credits = 1024u64;
+        let shuould_transmit = server.add_read_credits(credits);
+        assert_eq!(server.receive_window_shrink_debt, 0);
+        assert_eq!(server.local_max_data, prev_local_max_data + credits);
+        assert!(shuould_transmit.should_transmit());
+    }
+
+    #[test]
+    fn shrink_receive_window() {
+        let mut server = make(Side::Server);
+        let new_receive_window = 1024 * 512u32;
+        let prev_local_max_data = server.local_max_data;
+
+        // shrink the receive_winbow, local_max_data is not expected to be changed
+        let shrink_diff = server.receive_window - new_receive_window as u64;
+        let explanded = server.set_receive_window(new_receive_window.into());
+        assert!(!explanded);
+        assert_eq!(server.receive_window, new_receive_window as u64);
+        assert_eq!(server.local_max_data, prev_local_max_data);
+        assert_eq!(server.receive_window_shrink_debt, shrink_diff);
+        let prev_local_max_data = server.local_max_data;
+
+        // credit, local_max_data does not change as it is absorbed by receive_window_shrink_debt
+        let credits = 1024u64;
+        let expected_receive_window_shrink_debt = server.receive_window_shrink_debt - credits;
+        let shuould_transmit = server.add_read_credits(credits);
+        assert_eq!(
+            server.receive_window_shrink_debt,
+            expected_receive_window_shrink_debt
+        );
+        assert_eq!(server.local_max_data, prev_local_max_data);
+        assert!(!shuould_transmit.should_transmit());
+
+        // credit again, local_max_data does not change as it is absorbed by receive_window_shrink_debt
+        let expected_receive_window_shrink_debt = server.receive_window_shrink_debt - credits;
+        let shuould_transmit = server.add_read_credits(credits);
+        assert_eq!(
+            server.receive_window_shrink_debt,
+            expected_receive_window_shrink_debt
+        );
+        assert_eq!(server.local_max_data, prev_local_max_data);
+        assert!(!shuould_transmit.should_transmit());
```

Comment:
This looks like the exact same code ran twice. Convert into a loop of the form `for _ in 0..2` for better clarity?

Reply:
Done